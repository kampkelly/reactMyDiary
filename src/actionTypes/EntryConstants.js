export const NEW_ENTRY = 'NEW_ENTRY';
