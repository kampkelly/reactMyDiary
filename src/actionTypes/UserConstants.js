export const SIGNIN = 'SIGNIN';
export const SIGNUP = 'SIGNUP';
export const SIGNOUT = 'SIGNOUT';
export const PROFILE = 'PROFILE';
export const NOTIFICATION = 'NOTIFICATION';