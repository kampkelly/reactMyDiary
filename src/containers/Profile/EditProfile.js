import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import './EditProfile.scss';
import '../../styles/sidebar.scss';
import { ShowProfile, UpdateProfile } from '../../requests/UserRequests';
import { validateForm } from '../../helpers/validateForm';

/**
 * @class EditProfile
 * @extends {Component}
 */
class EditProfile extends Component {
  /**
   *Creates an instance of EditProfile.
   * @memberof EditProfile
   */
  constructor() {
    super();
    this.state = {
      editMode: false,
      dateOfBirth: '',
      email: '',
      fullName: '',
      mmessage: '',
      success: false,
      user: {}
    };
    this.change = this.change.bind(this);
    this.updateProfile = this.updateProfile.bind(this);
    this.goToHome = this.goToHome.bind(this);
  }

  change(e) {
    this.setState({ [e.target.name]: e.target.value });
  }

  componentDidMount() {
    document.querySelector('main').style.backgroundImage = "url('https://i.imgur.com/n4ttyU5.jpg')";
    this.props.ShowProfile();
    document.querySelector('body').insertAdjacentHTML('afterbegin', `<img src="http://res.cloudinary.com/ddfepbdqg/image/upload/v1543597369/Rolling.svg" id="loading"></img>`);
  }

  /**
   * @description - This method runs whenever redux state changes
   * @returns {object} state
   * @param {object} props
   * @param {object} state
   * @memberof EditProfile
   */
  static getDerivedStateFromProps(props, state) {
    if (props.user.email) {
      document.getElementById('loading').style.display = 'none';
    }
    if (state.editMode) {
      return {
        user: props.user,
        success: props.success
      };
    }
    if (!state.edit && props.user.dateofbirth) {
      return {
        editMode: true,
        dateOfBirth: props.user.dateofbirth,
        email: props.user.email,
        fullName: props.user.fullname,
        user: props.user,
        success: props.success
      };
    }
  }

  goToHome() {
    this.props.history.push('/dashboard');
  }

  /**
   *
   * @returns {object} state
   * @memberof EditProfile
   */
  updateProfile(e) {
    e.preventDefault();
    const errorMessage = 'One or more of the required fields are empty!';
    const { dateOfBirth, email, fullName } = this.state;
    const values = { email, fullName, dateOfBirth };
    const noneEmpty = validateForm(errorMessage, values, e);
    if (noneEmpty === true) {
      document.querySelector('body').insertAdjacentHTML('afterbegin', `<img src="http://res.cloudinary.com/ddfepbdqg/image/upload/v1543597369/Rolling.svg" id="loading"></img>`);
      this.props.UpdateProfile(email, fullName, dateOfBirth);
    }
  }

  /**
   * @description - This method renders the jsx for this component
   * @returns {jsx} - jsx
   * @memberof EditProfile
   */
  render() {
    const { dateOfBirth, email, fullName } = this.state;
    return (
      <div id="edit-profile">
         <div className="blue-forms form-cover text-center">
            <h3 className="underline title text-center">Edit My Profile</h3>
            <form className="edit-profile-form" id="edit-profile-form">
              <p className="danger-text form_error_text"><small></small></p>
              <p className="success-text form_success_text"><small></small></p>
              <div className="field grid-container">
                <div className="col-2-6 col-6-6-md col-6-6-xs">
                  <label className="">Full Name:</label>
                </div>
                <div className="col-4-6 col-6-6-md col-6-6-xs">
                  <input type="text" required="true" className="control-form" id="fullName" autoComplete="name" name="fullName" onChange={this.change} defaultValue={fullName} />
                </div>
              </div>
              <div className="field grid-container">
                <div className="col-2-6 col-6-6-md col-6-6-xs">
                  <label className="">Email Address:</label>
                </div>
                <div className="col-4-6 col-6-6-md col-6-6-xs">
                  <input type="email" required="true" className="control-form" id="email" autoComplete="email" name="email" onChange={this.change} defaultValue={email} />
                </div>
              </div>
              <div className="field grid-container">
                <div className="col-2-6 col-6-6-md col-6-6-xs">
                  <label className="">Date of Birth:</label>
                </div>
                <div className="col-4-6 col-6-6-md col-6-6-xs">
                  <input type="date" className="control-form" id="dateOfBirth" autoComplete="date" name="dateOfBirth" onChange={this.change} defaultValue={dateOfBirth} pattern="[a-z|A-Z|\d]{1,15}" />
                </div>
              </div>
              <div className="field grid-container">
                <div className="col-2-6 col-6-6-md col-6-6-xs"></div>
                <div className="col-2-6 col-6-6-md col-6-6-xs">
                  <button type="submit" className="submit-button button-white" onClick={this.updateProfile}>Update Profile</button>
                </div>
                <div className="col-2-6 col-6-6-md col-6-6-xs cancel-div">
                  <button type="submit" className="submit-button  button-cancel mt-4-xs mt-4-md" onClick={this.goToHome}>Cancel</button>
                </div>
                <div className="col-3-6 col-6-6-md col-6-6-xs">

                </div>
              </div>
            </form>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  user: state.user.user,
  success: state.user.success
});

EditProfile.propTypes = {
  ShowProfile: PropTypes.func,
  UpdateProfile: PropTypes.func,
  success: PropTypes.bool
};

export default connect(mapStateToProps, {
  ShowProfile, UpdateProfile
})(EditProfile);

