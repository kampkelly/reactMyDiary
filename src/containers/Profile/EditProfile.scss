$primary-color: #052F60;
$secondary-color: #DFAC2C;
$light-white-color: #8c8c8c;
$white-color: #fafafa;
$smartphone: "(min-width: 0px) and (max-width: 599px)";
$tablet: "(min-width: 600px) and (max-width: 900px)";

main {
	background-image: url("https://i.imgur.com/n4ttyU5.jpg");
    background-size: cover;
	background-position: center;
	background-repeat-x: no-repeat;
    background-repeat-y: no-repeat;
}
#edit-profile {
	padding-top: 20px;
    h3 {
        margin-bottom: 0;
        padding-top: 10px;
    }
    fieldset {
        border: none;
    }

    .form-cover {
		background: rgba(107, 101, 101, 0.3);
        margin: auto;
        .field {
			.control-form {
				width: 100%;
			}
            padding: 10px;
            border-radius: 10px;
            &:last-of-type {
                background: none;
            }
            button {
				width: 100%;
                &:hover {
					transition: 1s;
                }
            }
        }
    }
    [class*='col-'] {
      margin: auto;
    }
    .cancel-div {
      padding-right: 0px;
    }
	@media #{$tablet} {
		[class*='col-'] {
		  padding-right: 0px;
		}
    }
}
