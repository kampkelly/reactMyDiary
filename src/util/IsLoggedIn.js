export const isLoggedIn = () => {
  if (localStorage.getItem('diary_token') !== null) {
    return true;
  }
  return false;
};
