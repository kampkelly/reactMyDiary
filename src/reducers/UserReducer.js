import { SIGNIN, SIGNUP, SIGNOUT, PROFILE, NOTIFICATION, UPDATE_PROFILE } from '../actionTypes/UserConstants';
import { asyncActionName } from '../util/AsyncUtil';


const initialState = {
  isAuth: false,
  loading: false,
  message: '',
  success: false,
  failure: false,
  user: {}
};

const UserReducer = (state = initialState, action = {}) => {
  switch (action.type) {
    case asyncActionName(SIGNIN).loading:
      return {
        ...state,
        loading: action.payload,
      };
    case asyncActionName(SIGNIN).success:
      return {
        ...state,
        success: true, ...action.payload.user, isAuth: true
      };
    case asyncActionName(SIGNIN).failure:
      document.querySelector('.form_error_text').style.display = 'block';
      document.querySelector('.form_error_text small').textContent = action.payload.error;
      return {
        ...state,
      };
    case asyncActionName(SIGNUP).loading:
      return {
        ...state,
        loading: action.payload,
      };
    case asyncActionName(SIGNUP).success:
      return {
        ...state,
        success: action.payload,
      };
    case asyncActionName(SIGNUP).failure:
      document.querySelector('.form_error_text').style.display = 'block';
      document.querySelector('.form_error_text small').textContent = action.payload.error;
      return {
        ...state,
      };
    case asyncActionName(PROFILE).loading:
      return {
        ...state,
        loading: action.payload,
      };
    case asyncActionName(PROFILE).success:
      return {
        ...state,
        success: true, user: action.payload
      };
    case asyncActionName(PROFILE).failure:
      return {
        ...state,
      };
    case asyncActionName(UPDATE_PROFILE).loading:
      return {
        ...state,
        loading: action.payload,
      };
    case asyncActionName(UPDATE_PROFILE).success:
      return {
        ...state,
        message: action.payload.message, success: true, user: action.payload.user
      };
    case asyncActionName(UPDATE_PROFILE).failure:
      document.querySelector('.form_error_text').style.display = 'block';
			document.querySelector('.form_error_text small').textContent = data.message;
      return {
        ...state,
      };
    case asyncActionName(NOTIFICATION).loading:
      return {
        ...state,
        loading: action.payload,
      };
    case asyncActionName(NOTIFICATION).success:
      return {
        ...state,
        success: true, message: action.payload.message, user: action.payload.user
      };
    case asyncActionName(NOTIFICATION).failure:
      return {
        ...state,
      };
    case asyncActionName(SIGNOUT).success:
      return initialState;
    default:
      return state;
  }
};

export default userReducer;
