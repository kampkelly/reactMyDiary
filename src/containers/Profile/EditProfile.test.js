import React from 'react';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import EditProfile from './EditProfile';
import UserReducer from '../../reducers/UserReducer';
import { asyncActions } from '../../util/AsyncUtil';
import { UPDATE_PROFILE } from '../../actionTypes/UserConstants';

const mockStore = configureMockStore([thunk]);
const store = mockStore({
  user: {
    user: {},
    message: '',
    success: false
  },
  entry: {
    entries: []
  }
});

const props = {
  history: {
    push: jest.fn()
  },
  updateProfile: jest.fn(),
};

let component;
let myComponent;

describe('<EditProfile/>', () => {
  beforeEach(() => {
    component = shallow(
      <Provider store={store}>
        <EditProfile {...props} />
      </Provider>
    );
    myComponent = component.dive({ context: { store } }).dive();
  });
  it('should render without throwing an error', () => {
    expect(component).toMatchSnapshot();
  });
  it('should show profile', () => {
    const payload = {};
    const expectedAction = {
      type: 'UPDATE_PROFILE_SUCCESS',
      payload
    };
    expect(asyncActions(UPDATE_PROFILE).success(payload)).toEqual(expectedAction);
  });
  it('should dispatch an action when updating profile is loading', () => {
    const payload = false;
    const action = {
      type: 'UPDATE_PROFILE_LOADING',
      payload
    };
    const newState = UserReducer({}, action);
    expect(newState).toEqual({
      loading: payload
    });
  });
  it('should dispatch an action to update a profile', () => {
    const payload = {
      message: '',
      user: {}
    };
    const action = {
      type: 'UPDATE_PROFILE_SUCCESS',
      payload
    };
    const newState = UserReducer({}, action);
    expect(newState).toEqual({
      message: action.payload.message,
      user: action.payload.user,
      success: true,
    });
  });
  it('should dispatch an action when updating an entry fails', () => {
    document.body.innerHTML =
    '<div>' +
    '  <span id="loading" />' +
    '  <span class="form_error_text" /><small></small><span>' +
    '  <button id="button" />' +
    '</div>';
    const payload = {};
    const action = {
      type: 'UPDATE_PROFILE_FAILING',
      payload
    };
    const newState = UserReducer({}, action);
    expect(newState).toEqual({});
  });
  it('should update profile', () => {
    document.body.innerHTML =
    '<div>' +
    '  <span id="loading" />' +
    '  <span class="form_error_text" /><small></small><span>' +
    '  <button id="button" />' +
    '</div>';
    const input = myComponent.find('input#fullName');
    input.simulate('change', { target: { value: 'New name' } });
    myComponent.find('button.button-white').simulate('click', {
      preventDefault: () => {
      }
     });
  });
  it('should click cancel button', () => {
    myComponent.find('button.button-cancel').simulate('click');
  });
});
