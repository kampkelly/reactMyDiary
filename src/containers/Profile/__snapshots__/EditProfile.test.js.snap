// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditProfile/> should render without throwing an error 1`] = `
<Connect(EditProfile)
  history={
    Object {
      "push": [MockFunction],
    }
  }
  updateProfile={[MockFunction]}
/>
`;
